File: python/day05.py
from collections import namedtuple


Mapping = namedtuple("Mapping", "src dst length")


lines = open("../inputs/05.txt").read().strip().split("\n\n")
seeds = [int(x) for x in lines[0].split()[1:]]
stages = []
for m in lines[1:]:
    stages.append([])
    for mapping in m.split("\n")[1:]:
        dst_range_start, src_range_start, range_length = [int(x) for x in mapping.split()]
        stages[-1].append(Mapping(src=src_range_start, dst=dst_range_start, length=range_length))


def overlap(b0, len0, b1, len1):
    if b0 > b1:
        b0, b1, len0, len1 = b1, b0, len1, len0
    b, e = max(b0, b1), min(b0 + len0, b1 + len1)
    if e > b:
        return b, e - b
    else:
        return None


def expand_range(begin, length, stage_id):
    if stage_id == len(stages):
        return begin
    matches = []
    for mapping in stages[stage_id]:
        o = overlap(begin, length, mapping.src, mapping.length)
        if o:
            overlap_begin, overlap_length = o
            overlap_begin += mapping.dst - mapping.src
            matches.append(expand_range(overlap_begin, overlap_length, stage_id + 1))
    if not matches:
        return expand_range(begin, length, stage_id + 1)
    return min(matches)


print(min(expand_range(s, 1, 0) for s in seeds))
print(min(expand_range(s, l, 0) for s, l in zip(seeds[0::2], seeds[1::2])))

File: cpp/day05.cpp
#include <algorithm>
#include <chrono>
#include <fstream>
#include <iostream>
#include <optional>
#include <sstream>
#include <vector>

struct Mapping
{
    int64_t src, dst, len;
};

[[nodiscard]] std::optional<std::pair<int64_t, int64_t>> overlap(int64_t b0, int64_t len0, int64_t b1, int64_t len1)
{
    if (b0 > b1) {
        std::swap(b0, b1);
        std::swap(len0, len1);
    }
    auto const b = std::max(b0, b1);
    auto const e = std::min(b0 + len0, b1 + len1);
    if (e > b) {
        auto pair = std::make_pair(b, e - b);
        return {pair};
    }
    return {};
}

using Stages = std::vector<std::vector<Mapping>>;

[[nodiscard]] int64_t expand_range(Stages const& stages, int64_t begin, int64_t const len, size_t const stage_id)
{
    if (stage_id == stages.size()) {
        return begin;
    }
    int64_t min = std::numeric_limits<int64_t>::max();
    std::vector<std::pair<int64_t, int64_t>> match_ranges;
    for (auto const& mapping : stages[stage_id]) {
        auto maybe_overlap = overlap(begin, len, mapping.src, mapping.len);
        if (maybe_overlap.has_value()) {
            auto [overlap_begin, overlap_len] = maybe_overlap.value();
            match_ranges.push_back(maybe_overlap.value());
            overlap_begin += mapping.dst - mapping.src;
            auto loc = expand_range(stages, overlap_begin, overlap_len, stage_id + 1);
            if (loc < min) {
                min = loc;
            }
        }
    }
    if (match_ranges.empty()) {
        return expand_range(stages, begin, len, stage_id + 1);
    }

    for (auto const [match_x, match_len] : match_ranges) {
        if (begin < match_x) {
            auto const loc = expand_range(stages, begin, match_x - begin, stage_id + 1);
            if (loc < min) {
                min = loc;
            }
        }
        begin = match_x + match_len;
    }
    return min;
}

int main()
{
    auto begin = std::chrono::high_resolution_clock::now();
    std::vector<int64_t> seeds;
    std::ifstream f{"../inputs/05.txt"};
    Stages stages{7};

    std::string line;
    std::getline(f, line);
    line = line.substr(7);
    std::istringstream iss(line);
    int64_t i;
    while (iss >> i) {
        seeds.push_back(i);
    }
    std::getline(f, line);

    for (auto map_id = 0; map_id < 7; ++map_id) {
        auto& stage = stages.at(map_id);
        std::getline(f, line);
        while (!f.eof()) {
            std::getline(f, line);
            if (line.length() < 5) break;
            std::istringstream iss(line);
            int64_t dst, src, len;
            iss >> dst >> src >> len;
            stage.push_back(Mapping{src, dst, len});
        }
        std::sort(stage.begin(), stage.end(), [](auto const& a, auto const& b) { return a.src < b.src; });
    }

    {
        int64_t min = std::numeric_limits<int64_t>::max();
        for (auto s : seeds) {
            auto loc = expand_range(stages, s, 1, 0);
            if (loc < min) min = loc;
        }
        std::cout << min << '\n';
    }

    {
        int64_t min = std::numeric_limits<int64_t>::max();
        for (auto i = 0; i < seeds.size(); i += 2) {
            auto loc = expand_range(stages, seeds[i], seeds[i + 1], 0);
            if (loc < min) min = loc;
        }
        std::cout << min << '\n';
    }
    std::cout << (std::chrono::high_resolution_clock::now() - begin).count() << " ns\n";
}